In [618]:
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sn 
import numpy as np
import seaborn as sns

#### Open GTDBtk results (bac120andAr122.summary.tsv)

In [619]:
f = open("data56_related_files/gtdbtk.bac_ar_summary_relevant_columns.tsv", "r")
files = f.read()

#### Create list and dataframe 

In [620]:
fileList = files.split("\n")

In [621]:
df = pd.DataFrame(fileList)

#### Cleaning df and splitting string to columns

In [622]:
df.columns = ['temp']
# user_genome	classification	fastani_reference	fastani_reference_radius	fastani_taxonomy	
# fastani_ani	fastani_af	closest_placement_reference	closest_placement_radius	closest_placement_taxonomy	
# closest_placement_ani	closest_placement_af	pplacer_taxonomy	classification_method	note	
# other_related_references(genome_id,species_name,radius,ANI,AF)	msa_percent	translation_table	red_value
# warnings

df[['user_genome','classification', 'closest_placement_taxonomy',
    'msa_percent']] = df.temp.str.split("\t",expand=True)


In [623]:
df.head(3)

temp  user_genome  \
0  user_genome\tclassification\tclosest_placement...  user_genome   
1  S10C1032\td__Bacteria;p__Bacteroidota;c__Bacte...     S10C1032   
2  S10C1039\td__Bacteria;p__Firmicutes;c__Bacilli...     S10C1039   

                                      classification  \
0                                     classification   
1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
2  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   

                          closest_placement_taxonomy  msa_percent  
0                         closest_placement_taxonomy  msa_percent  
1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...        33.75  
2  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...        94.58

In [624]:
# drop first row because we get double headers 
df = df.drop(0)

# drop the last row since it is empty
last_row = len(df)
df = df.drop(df.index[last_row-1]) 

In [625]:
# drop first coulmn that is the original combined string
df = df.iloc[: , 1:]

In [626]:
# Because archea and bacteria were concatinated the header line is present further down where the archea file starts. 
df.drop(index=df[df['user_genome'] == 'user_genome'].index, inplace=True)

In [627]:
df

user_genome                                     classification  \
1       S10C1032  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
2       S10C1039  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
3       S10C1113  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
4        S10C119  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
5       S10C1295  d__Bacteria;p__Campylobacterota;c__Campylobact...   
...          ...                                                ...   
2378      S9C962  d__Bacteria;p__Campylobacterota;c__Campylobact...   
2379       S9C99  d__Bacteria;p__Firmicutes_C;c__Negativicutes;o...   
2381    S23C1277  d__Archaea;p__Methanobacteriota;c__Methanobact...   
2382    S34C1277  d__Archaea;p__Methanobacteriota;c__Methanobact...   
2383    S53C3468  d__Archaea;p__Methanobacteriota;c__Methanobact...   

                             closest_placement_taxonomy msa_percent  
1     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       33.75  
2     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       94.58  
3     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       96.07  
4     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       67.12  
5     d__Bacteria;p__Campylobacterota;c__Campylobact...       44.93  
...                                                 ...         ...  
2378  d__Bacteria;p__Campylobacterota;c__Campylobact...       93.17  
2379  d__Bacteria;p__Firmicutes_C;c__Negativicutes;o...       51.82  
2381  d__Archaea;p__Methanobacteriota;c__Methanobact...       74.32  
2382  d__Archaea;p__Methanobacteriota;c__Methanobact...       41.32  
2383  d__Archaea;p__Methanobacteriota;c__Methanobact...       56.79  

[2382 rows x 4 columns]

In [628]:
#splitting classification to columns based on tax-rank
df[['Domain','Phylum', 'Class', 'Order', 'Family', 
    'Genus', 'Species']] = df.classification.str.split(";",expand=True)
#splitting user_genome ID to sample and cluster
df[['sample','cluster']] = df.user_genome.str.split("C",expand=True)

In [629]:
df

user_genome                                     classification  \
1       S10C1032  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
2       S10C1039  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
3       S10C1113  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
4        S10C119  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
5       S10C1295  d__Bacteria;p__Campylobacterota;c__Campylobact...   
...          ...                                                ...   
2378      S9C962  d__Bacteria;p__Campylobacterota;c__Campylobact...   
2379       S9C99  d__Bacteria;p__Firmicutes_C;c__Negativicutes;o...   
2381    S23C1277  d__Archaea;p__Methanobacteriota;c__Methanobact...   
2382    S34C1277  d__Archaea;p__Methanobacteriota;c__Methanobact...   
2383    S53C3468  d__Archaea;p__Methanobacteriota;c__Methanobact...   

                             closest_placement_taxonomy msa_percent  \
1     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       33.75   
2     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       94.58   
3     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       96.07   
4     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       67.12   
5     d__Bacteria;p__Campylobacterota;c__Campylobact...       44.93   
...                                                 ...         ...   
2378  d__Bacteria;p__Campylobacterota;c__Campylobact...       93.17   
2379  d__Bacteria;p__Firmicutes_C;c__Negativicutes;o...       51.82   
2381  d__Archaea;p__Methanobacteriota;c__Methanobact...       74.32   
2382  d__Archaea;p__Methanobacteriota;c__Methanobact...       41.32   
2383  d__Archaea;p__Methanobacteriota;c__Methanobact...       56.79   

           Domain                Phylum               Class  \
1     d__Bacteria       p__Bacteroidota      c__Bacteroidia   
2     d__Bacteria         p__Firmicutes          c__Bacilli   
3     d__Bacteria         p__Firmicutes          c__Bacilli   
4     d__Bacteria         p__Firmicutes          c__Bacilli   
5     d__Bacteria   p__Campylobacterota  c__Campylobacteria   
...           ...                   ...                 ...   
2378  d__Bacteria   p__Campylobacterota  c__Campylobacteria   
2379  d__Bacteria       p__Firmicutes_C    c__Negativicutes   
2381   d__Archaea  p__Methanobacteriota  c__Methanobacteria   
2382   d__Archaea  p__Methanobacteriota  c__Methanobacteria   
2383   d__Archaea  p__Methanobacteriota  c__Methanobacteria   

                      Order                  Family                     Genus  \
1          o__Bacteroidales       f__Bacteroidaceae            g__Phocaeicola   
2        o__Lactobacillales     f__Lactobacillaceae    g__Limosilactobacillus   
3        o__Lactobacillales     f__Lactobacillaceae    g__Limosilactobacillus   
4        o__Lactobacillales     f__Lactobacillaceae              g__Weissella   
5      o__Campylobacterales    f__Helicobacteraceae         g__Helicobacter_D   
...                     ...                     ...                       ...   
2378   o__Campylobacterales    f__Helicobacteraceae         g__Helicobacter_G   
2379   o__Acidaminococcales   f__Acidaminococcaceae  g__Phascolarctobacterium   
2381  o__Methanobacteriales  f__Methanobacteriaceae   g__Methanobrevibacter_A   
2382  o__Methanobacteriales  f__Methanobacteriaceae   g__Methanobrevibacter_A   
2383  o__Methanobacteriales  f__Methanobacteriaceae   g__Methanobrevibacter_A   

                                   Species sample cluster  
1                     s__Phocaeicola dorei    S10    1032  
2         s__Limosilactobacillus ingluviei    S10    1039  
3              s__Limosilactobacillus oris    S10    1113  
4               s__Weissella thailandensis    S10     119  
5               s__Helicobacter_D pullorum    S10    1295  
...                                    ...    ...     ...  
2378                                   s__     S9     962  
2379  s__Phascolarctobacterium sp000436095     S9      99  
2381        s__Methanobreviba

In [630]:
df.cluster = df.cluster.astype(int)

#### Function to check if the bins in the same cluster has recieved the same taxonomical assignation

In [631]:
def CheckingIdenticalTaxInCluster(df):

    sortedByCluster = df.sort_values(by=['cluster'])
    rows = sortedByCluster.iterrows()
    _, nextRow = next(rows)  # take first item from row_iterator
    countSpecies, countGenus, countFamily  = 0, 0, 0 

    for i, row in rows:
        if (ClusterValuesIsEqual(row, nextRow)): 
            if not (SpeciesNotDefined(row, nextRow)):
                if (IsdifferentSpecieces(row, nextRow)):
                    ShowDifferenceInTaxRank(row, nextRow)
                    print('species differ')
                    countSpecies += 1
            if (SpeciesNotDefined(row, nextRow)):
                if not (GenusNotDefined(row, nextRow)):
                    if (IsdifferentGenus(row, nextRow)):
                        ShowDifferenceInTaxRank(row, nextRow)
                        print('Genus differ')
                        countGenus += 1
                if (GenusNotDefined(row, nextRow)):
                    if (IsFamilyDifferent(row, nextRow)):
                        ShowDifferenceInTaxRank(row, nextRow)
                        print('family differ')      
                        countFamily += 1
        nextRow = row
    
    print('number of times where species differ:', countSpecies) 
    print('number of times where genus differ:', countGenus)
    print('number of times where family differ:', countFamily)    

In [721]:
# a test to see if I can get it to one bin against all in the same cluster. 
def CheckingIdenticalTaxInCluster_all(df):

    df.sort_values(by=['cluster'], inplace=True)
#     rows = sortedByCluster.iterrows()
#     _, nextRow = next(rows)  # take previous item from rows
#     countSpecies, countGenus, countFamily  = 0, 0, 0 
    
    # 1. iterate though clusers
    
    
    for cluster in range(df['cluster'].min(), df['cluster'].max()): # consider changing to groupby cluster
        c = df.loc[df['cluster'] == cluster]
        
        # value in cluster is unique
        
        countSpecies = 0
        unique = c['Species'].drop_duplicates()
#         print(unique == 's__')
        
       
        if (unique.count() > 1 ) :

            if not (unique.values.any() == 's__'): # dette virker ikke - kigger kun på det første element
                
                print(unique.values) 
                countSpecies += 1

#         c.loc[c['Species'].unique()]
        
        # more than one unique species in the cluster - then print
        
        
        
              

#     for i, row in rows:
#         clusterValue = GetClusterValue(row)
#         if (ClusterValuesIsEqual(row, nextRow)): 
#             if not (SpeciesNotDefined(row, nextRow)):
#                 if (IsdifferentSpecieces(row, nextRow)):
#                     ShowDifferenceInTaxRank(row, nextRow)
#                     print('species differ')
#                     countSpecies += 1
#             if (SpeciesNotDefined(row, nextRow)):
#                 if not (GenusNotDefined(row, nextRow)):
#                     if (IsdifferentGenus(row, nextRow)):
#                         ShowDifferenceInTaxRank(row, nextRow)
#                         print('Genus differ')
#                         countGenus += 1
#                 if (GenusNotDefined(row, nextRow)):
#                     if (IsFamilyDifferent(row, nextRow)):
#                         ShowDifferenceInTaxRank(row, nextRow)
#                         print('family differ')      
#                         countFamily += 1
            
        #nextRow = row
    
    print('number of times where species differ:', countSpecies) 
    print('number of times where genus differ:', countGenus)
    print('number of times where family differ:', countFamily)  
CheckingIdenticalTaxInCluster_all(df)

['s__HGM11588 sp900754755' 's__']
['s__Corynebacterium pollutisoli' 's__Corynebacterium xerosis']
['s__Escherichia flexneri' 's__Escherichia coli_D']
['s__Mediterraneibacter sp002161355' 's__']
['s__Enterococcus_E cecorum' 's__']
['s__Ligilactobacillus aviarius' 's__']
['s__AM51-8 sp900761925' 's__']
['s__UMGS1004 sp900548845' 's__']
['s__Sellimonas sp002161525' 's__']
number of times where species differ: 0


NameError: name 'countGenus' is not defined

In [638]:
df.loc[df['cluster'] == 1]

user_genome                                     classification  \
727        S27C1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
1868       S50C1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
659        S26C1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
1543       S45C1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   

                             closest_placement_taxonomy msa_percent  \
727                                                 N/A       11.12   
1868  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       15.39   
659                                                 N/A        19.0   
1543                                                N/A       10.56   

           Domain           Phylum           Class             Order  \
727   d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
1868  d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
659   d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
1543  d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   

                   Family           Genus Species sample  cluster  
727   f__Coprobacteraceae             g__     s__    S27        1  
1868   f__Barnesiellaceae  g__Barnesiella     s__    S50        1  
659   f__Coprobacteraceae             g__     s__    S26        1  
1543  f__Coprobacteraceae             g__     s__    S45        1

In [560]:
# Måske noget alla dette kan bruges til noget: nok ikke :( 
def function(x):
    i = x.query(
        'area == 12 and color == "red" and shape == "circle"'
    )
    j = x.query(
        'area == 13 and color == "green" and shape == "square"'
    )
    return x if not (i.empty or j.empty) else x[x.num == x.num.max()].head(1)

In [561]:
def GetClusterValue(row):
    return row['cluster']

In [562]:
def ClusterValuesIsEqual(row, nextRow):
    return row['cluster'] == nextRow['cluster']

In [563]:
def IsdifferentSpecieces(row, nextRow):
    return row['Species'] != nextRow['Species']

In [564]:
def ShowDifferenceInTaxRank(row, nextRow):
    print(row['user_genome'], row['classification'])
    print(nextRow['user_genome'], nextRow['classification'])

In [565]:
def SpeciesNotDefined(row, nextRow):
    return row['Species'] == 's__' or nextRow['Species'] == 's__'

In [566]:
def IsdifferentGenus(row, nextRow):
    return row['Genus'] != nextRow['Genus'] 

In [567]:
def GenusNotDefined(row, nextRow):
    return row['Genus'] == 'g__' or nextRow['Genus'] == 'g__'

In [568]:
def IsFamilyDifferent(row, nextRow):
    return row['Family'] != nextRow['Family'] 

In [569]:
CheckingIdenticalTaxInCluster_all(df)

S50C1 d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Barnesiellaceae;g__Barnesiella;s__
S27C1 d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Coprobacteraceae;g__;s__
family differ
number of times where species differ: 0
number of times where genus differ: 0
number of times where family differ: 1


In [580]:
df['cluster'].min()

1

In [437]:
# Where family differ: 
df[df["cluster"] == 1]
# One cluster leads to 3 times where species differs. (we only get 2? - why no comparison S50 vs S45)
# we need to redo CheckingIdenticalTaxInCluster funciton: 
#it only compares 1 to 2, 2 to 3 and not all to all within a cluster - it would be enough to compare 
# 1 to all and not all to all. 

user_genome                                     classification  \
659        S26C1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
727        S27C1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
1543       S45C1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
1868       S50C1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   

                             closest_placement_taxonomy msa_percent  \
659                                                 N/A        19.0   
727                                                 N/A       11.12   
1543                                                N/A       10.56   
1868  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       15.39   

           Domain           Phylum           Class             Order  \
659   d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
727   d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
1543  d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
1868  d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   

                   Family           Genus Species sample  cluster  
659   f__Coprobacteraceae             g__     s__    S26        1  
727   f__Coprobacteraceae             g__     s__    S27        1  
1543  f__Coprobacteraceae             g__     s__    S45        1  
1868   f__Barnesiellaceae  g__Barnesiella     s__    S50        1

Number of undefined families:

In [363]:
df[df["Family"] == 'f__'].count()

user_genome                   0
classification                0
closest_placement_taxonomy    0
msa_percent                   0
Domain                        0
Phylum                        0
Class                         0
Order                         0
Family                        0
Genus                         0
Species                       0
sample                        0
cluster                       0
dtype: int64

In [368]:
df[df["Genus"] == 'g__'].count()

user_genome                   108
classification                108
closest_placement_taxonomy    108
msa_percent                   108
Domain                        108
Phylum                        108
Class                         108
Order                         108
Family                        108
Genus                         108
Species                       108
sample                        108
cluster                       108
dtype: int64

In [370]:
df[df["Species"] == 's__'].count()

user_genome                   1093
classification                1093
closest_placement_taxonomy    1093
msa_percent                   1093
Domain                        1093
Phylum                        1093
Class                         1093
Order                         1093
Family                        1093
Genus                         1093
Species                       1093
sample                        1093
cluster                       1093
dtype: int64

In [376]:
df.nunique()

user_genome                   2382
classification                 467
closest_placement_taxonomy     397
msa_percent                   1786
Domain                           2
Phylum                          17
Class                           21
Order                           52
Family                         100
Genus                          246
Species                        269
sample                          56
cluster                       1067
dtype: int64

### Overview
Out of 2382 bins/rows 

* 1093 species undefined 
* 108 genus undefined 
* 0 families undefined

Out of 1067 clusters 
* 3 times did species differ within two clusters 
* 6 times did genus differ within 4 different clusters
* 2 times did families differ within the same cluster C1. 


### Lactobacillus/Alistipes Bins
Generating dataframes with only Lactobacillus or Alistipes bins


In [415]:
Lactobacillus_df = df[df["Genus"] == 'g__Lactobacillus']
Lactobacillaceae_df =df[df["Family"] == 'f__Lactobacillaceae']

In [416]:
Lactobacillus_df = pd.DataFrame(Lactobacillus_df)
Lactobacillaceae_df = pd.DataFrame(Lactobacillaceae_df)

In [409]:
Alistipes_df = df[df["Genus"].isin(['g__Alistipes','g__Alistipes_A'])]
Alistipes_df = pd.DataFrame(Alistipes_df)

In [413]:
Alistipes_df

user_genome                                     classification  \
7      S10C15900  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
34        S10C90  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
71        S11C90  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
131      S12C685  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
138       S12C90  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
...          ...                                                ...   
2294     S8C3336  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
2319      S8C674  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
2326       S8C90  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
2331     S9C1094  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   
2336    S9C14905  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   

                             closest_placement_taxonomy msa_percent  \
7     d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       95.37   
34    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       14.08   
71    d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       90.49   
131   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       14.91   
138   d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       20.31   
...                                                 ...         ...   
2294  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       13.96   
2319  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       78.28   
2326  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       93.15   
2331  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       69.35   
2336  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...       93.41   

           Domain           Phylum           Class             Order  \
7     d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
34    d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
71    d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
131   d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
138   d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
...           ...              ...             ...               ...   
2294  d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
2319  d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
2326  d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
2331  d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   
2336  d__Bacteria  p__Bacteroidota  c__Bacteroidia  o__Bacteroidales   

                Family         Genus                   Species sample  cluster  
7     f__Rikenellaceae  g__Alistipes  s__Alistipes sp000434235    S10    15900  
34    f__Rikenellaceae  g__Alistipes  s__Alistipes sp900021155    S10       90  
71    f__Rikenellaceae  g__Alistipes  s__Alistipes sp900021155    S11       90  
131   f__Rikenellaceae  g__Alistipes   s__Alistipes finegoldii    S12      685  
138   f__Rikenellaceae  g__Alistipes  s__Alistipes sp900021155    S12       90  
...                ...           ...                       ...    ...      ...  
2294  f__Rikenellaceae  g__Alistipes                       s__     S8     3336  
2319  f__Rikenellaceae  g__Alistipes       s__Alistipes dispar     S8      674  
2326  f__Rikenellaceae  g__Alistipes  s__Alistipes sp900021155     S8       90  
2331  f__Rikenellaceae  g__Alistipes  s__Alistipes sp002161445     S9     1094  
2336  f__Rikenellaceae  g__Alistipes  s__Alistipes sp900546065     S9    14905  

[99 rows x 13 columns]

In [394]:
Lactobacillus_df.to_csv('data56_related_files/GTDB_Lactobacillus.csv', index = False)
Alistipes_df.to_csv('data56_related_files/GTDB_Alistipes.csv', index = False)

In [417]:
Lactobacillaceae_df.to_csv('data56_related_files/GTDB_Lactobacillaceae.csv', index = False)

In [414]:
Lactobacillus_df.head(5)

user_genome                                     classification  \
13     S10C205  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
20     S10C362  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
28     S10C569  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
49     S11C205  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
61     S11C569  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   

                           closest_placement_taxonomy msa_percent  \
13  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       84.32   
20  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       70.24   
28  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...        95.0   
49  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...        72.6   
61  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       84.93   

         Domain         Phylum       Class               Order  \
13  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
20  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
28  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
49  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
61  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   

                 Family             Genus                      Species sample  \
13  f__Lactobacillaceae  g__Lactobacillus   s__Lactobacillus crispatus    S10   
20  f__Lactobacillaceae  g__Lactobacillus  s__Lactobacillus gallinarum    S10   
28  f__Lactobacillaceae  g__Lactobacillus   s__Lactobacillus johnsonii    S10   
49  f__Lactobacillaceae  g__Lactobacillus   s__Lactobacillus crispatus    S11   
61  f__Lactobacillaceae  g__Lactobacillus   s__Lactobacillus johnsonii    S11   

    cluster  
13      205  
20      362  
28      569  
49      205  
61      569

In [418]:
Lactobacillaceae_df

user_genome                                     classification  \
2       S10C1039  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
3       S10C1113  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
4        S10C119  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
6       S10C1297  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
8       S10C1752  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
...          ...                                                ...   
2358      S9C362  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
2359     S9C3913  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
2361      S9C442  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
2363     S9C4469  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
2369      S9C569  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   

                             closest_placement_taxonomy msa_percent  \
2     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       94.58   
3     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       96.07   
4     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       67.12   
6     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       93.47   
8     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       85.09   
...                                                 ...         ...   
2358  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       88.76   
2359                                                N/A       92.44   
2361  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       15.64   
2363  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       18.78   
2369  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       69.92   

           Domain         Phylum       Class               Order  \
2     d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
3     d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
4     d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
6     d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
8     d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
...           ...            ...         ...                 ...   
2358  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
2359  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
2361  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
2363  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
2369  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   

                   Family                   Genus  \
2     f__Lactobacillaceae  g__Limosilactobacillus   
3     f__Lactobacillaceae  g__Limosilactobacillus   
4     f__Lactobacillaceae            g__Weissella   
6     f__Lactobacillaceae    g__Ligilactobacillus   
8     f__Lactobacillaceae    g__Ligilactobacillus   
...                   ...                     ...   
2358  f__Lactobacillaceae        g__Lactobacillus   
2359  f__Lactobacillaceae    g__Ligilactobacillus   
2361  f__Lactobacillaceae  g__Limosilactobacillus   
2363  f__Lactobacillaceae    g__Ligilactobacillus   
2369  f__Lactobacillaceae        g__Lactobacillus   

                               Species sample  cluster  
2     s__Limosilactobacillus ingluviei    S10     1039  
3          s__Limosilactobacillus oris    S10     1113  
4           s__Weissella thailandensis    S10      119  
6          s__Ligilactobacillus agilis    S10     1297  
8      s__Ligilactobacillus salivarius    S10     1752  
...                                ...    ...      ...  
2358       s__Lactobacillus gallinarum     S9      362  
2359                               s__     S9     3913  
2361  s__Limosilactobacillus reuteri_E     S9      442  
2363     s__Ligilactobacillus aviarius     S9     4469  
2369        s__Lactobacillus johnsonii     S9      569  

[344 rows x 13 columns]

#### Lactobacillus from NCBI
Multiple GTDB genus are used to represent the Lactobacillus genus from NCBI. The NCBI_Lactobacillus_df should contain all bins ranked as a NCBI Lactobacillus.  

In [548]:
NCBI_Lactobacillus_df = pd.read_csv('GTDB_ref_genomes/Lactobacillus_representative_ref_genomes.tsv')

In [549]:
NCBI_Lactobacillus_df.head(5)

ID\t"NCBI Organism Name"\t"NCBI Taxonomy"\t"GTDB Taxonomy"\t"GTDB species representative"\t"NCBI type material"
0  GCA_000615245.1\t"Lactobacillus rhamnosus DSM ...                                                             
1  GCA_001311425.1\t"Lactobacillus kisonensis DSM...                                                             
2  GCA_001311745.1\t"Lactobacillus kefiri DSM 205...                                                             
3  GCA_001311765.1\t"Lactobacillus equicursoris D...                                                             
4  GCA_001435375.1\t"Lactobacillus aviarius subsp...

In [550]:
NCBI_Lactobacillus_df.columns = ['temp']
# user_genome	classification	fastani_reference	fastani_reference_radius	fastani_taxonomy	
# fastani_ani	fastani_af	closest_placement_reference	closest_placement_radius	closest_placement_taxonomy	
# closest_placement_ani	closest_placement_af	pplacer_taxonomy	classification_method	note	
# other_related_references(genome_id,species_name,radius,ANI,AF)	msa_percent	translation_table	red_value
# warnings

NCBI_Lactobacillus_df[['ID','NCBI_Organism_Name', 'NCBI_Taxonomy', 'GTDB_Taxonomy', 
    'GTDB_species_representative', 'NCBI_type_material']] = NCBI_Lactobacillus_df.temp.str.split("\t",expand=True)

NCBI_Lactobacillus_df = NCBI_Lactobacillus_df.iloc[: , 1:]

In [551]:
NCBI_Lactobacillus_df.head(5)

ID                                 NCBI_Organism_Name  \
0  GCA_000615245.1  "Lactobacillus rhamnosus DSM 20021 = JCM 1136 ...   
1  GCA_001311425.1   "Lactobacillus kisonensis DSM 19906 = JCM 15041"   
2  GCA_001311745.1        "Lactobacillus kefiri DSM 20587 = JCM 5818"   
3  GCA_001311765.1  "Lactobacillus equicursoris DSM 19284 = JCM 14...   
4  GCA_001435375.1  "Lactobacillus aviarius subsp. araffinosus DSM...   

                                       NCBI_Taxonomy  \
0  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
1  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
2  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
3  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
4  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   

                                       GTDB_Taxonomy  \
0  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
1  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
2  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
3  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
4  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   

  GTDB_species_representative NCBI_type_material  
0                       "yes"              "yes"  
1                       "yes"              "yes"  
2                       "yes"              "yes"  
3                       "yes"              "yes"  
4                       "yes"                 ""

In [552]:
NCBI_Lactobacillus_df[['Domain','Phylum', 'Class', 'Order', 
    'Family', 'Genus', 'Species']] = NCBI_Lactobacillus_df.GTDB_Taxonomy.str.split("; ",expand=True)



In [553]:
NCBI_Lactobacillus_df.head(5)

ID                                 NCBI_Organism_Name  \
0  GCA_000615245.1  "Lactobacillus rhamnosus DSM 20021 = JCM 1136 ...   
1  GCA_001311425.1   "Lactobacillus kisonensis DSM 19906 = JCM 15041"   
2  GCA_001311745.1        "Lactobacillus kefiri DSM 20587 = JCM 5818"   
3  GCA_001311765.1  "Lactobacillus equicursoris DSM 19284 = JCM 14...   
4  GCA_001435375.1  "Lactobacillus aviarius subsp. araffinosus DSM...   

                                       NCBI_Taxonomy  \
0  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
1  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
2  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
3  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
4  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   

                                       GTDB_Taxonomy  \
0  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
1  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
2  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
3  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   
4  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   

  GTDB_species_representative NCBI_type_material        Domain         Phylum  \
0                       "yes"              "yes"  "d__Bacteria  p__Firmicutes   
1                       "yes"              "yes"  "d__Bacteria  p__Firmicutes   
2                       "yes"              "yes"  "d__Bacteria  p__Firmicutes   
3                       "yes"              "yes"  "d__Bacteria  p__Firmicutes   
4                       "yes"                 ""  "d__Bacteria  p__Firmicutes   

        Class               Order               Family                  Genus  \
0  c__Bacilli  o__Lactobacillales  f__Lactobacillaceae  g__Lacticaseibacillus   
1  c__Bacilli  o__Lactobacillales  f__Lactobacillaceae  g__Lentilactobacillus   
2  c__Bacilli  o__Lactobacillales  f__Lactobacillaceae  g__Lentilactobacillus   
3  c__Bacilli  o__Lactobacillales  f__Lactobacillaceae       g__Lactobacillus   
4  c__Bacilli  o__Lactobacillales  f__Lactobacillaceae   g__Ligilactobacillus   

                             Species  
0   s__Lacticaseibacillus rhamnosus"  
1  s__Lentilactobacillus kisonensis"  
2      s__Lentilactobacillus kefiri"  
3     s__Lactobacillus equicursoris"  
4  s__Ligilactobacillus araffinosus"

In [558]:
NCBI_Lactobacillus_df["Family"].unique()

array(['f__Lactobacillaceae', 'f__Lachnospiraceae'], dtype=object)

In [557]:
NCBI_Lactobacillus_df[NCBI_Lactobacillus_df["Family"] == 'f__Lachnospiraceae']

ID       NCBI_Organism_Name  \
10  GCA_900112995.1  "Lactobacillus rogosae"   

                                        NCBI_Taxonomy  \
10  "d__Bacteria; p__Firmicutes; c__Bacilli; o__La...   

                                        GTDB_Taxonomy  \
10  "d__Bacteria; p__Firmicutes_A; c__Clostridia; ...   

   GTDB_species_representative NCBI_type_material        Domain  \
10                       "yes"                 ""  "d__Bacteria   

             Phylum          Class              Order              Family  \
10  p__Firmicutes_A  c__Clostridia  o__Lachnospirales  f__Lachnospiraceae   

             Genus                    Species  
10  g__Lachnospira  s__Lachnospira rogosae_A"

In [555]:
NCBI_Lactobacillus_df["Genus"].unique()

array(['g__Lacticaseibacillus', 'g__Lentilactobacillus',
       'g__Lactobacillus', 'g__Ligilactobacillus',
       'g__Furfurilactobacillus', 'g__Levilactobacillus',
       'g__Lachnospira', 'g__Limosilactobacillus',
       'g__Schleiferilactobacillus', 'g__Lactiplantibacillus',
       'g__Latilactobacillus', 'g__Companilactobacillus',
       'g__Secundilactobacillus', 'g__Paucilactobacillus',
       'g__Bombilactobacillus', 'g__Apilactobacillus',
       'g__Fructilactobacillus', 'g__Loigolactobacillus', 'g__Dellaglioa',
       'g__Lapidilactobacillus', 'g__Agrilactobacillus', 'g__Holzapfelia',
       'g__Paralactobacillus', 'g__Amylolactobacillus', 'g__RHOV01'],
      dtype=object)

In [556]:
NCBI_Lactobacillus_df["Genus"].nunique()

25

In [571]:
NCBI_genuses = NCBI_Lactobacillus_df["Genus"].unique()

In [572]:
NCBI_Lactobacillus_df = df[df["Genus"].isin(NCBI_genuses)]

In [573]:
NCBI_Lactobacillus_df

user_genome                                     classification  \
2       S10C1039  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
3       S10C1113  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
6       S10C1297  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
8       S10C1752  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
12       S10C188  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
...          ...                                                ...   
2358      S9C362  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
2359     S9C3913  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
2361      S9C442  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
2363     S9C4469  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   
2369      S9C569  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...   

                             closest_placement_taxonomy msa_percent  \
2     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       94.58   
3     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       96.07   
6     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       93.47   
8     d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       85.09   
12    d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...        94.9   
...                                                 ...         ...   
2358  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       88.76   
2359                                                N/A       92.44   
2361  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       15.64   
2363  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       18.78   
2369  d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactob...       69.92   

           Domain         Phylum       Class               Order  \
2     d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
3     d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
6     d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
8     d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
12    d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
...           ...            ...         ...                 ...   
2358  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
2359  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
2361  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
2363  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   
2369  d__Bacteria  p__Firmicutes  c__Bacilli  o__Lactobacillales   

                   Family                   Genus  \
2     f__Lactobacillaceae  g__Limosilactobacillus   
3     f__Lactobacillaceae  g__Limosilactobacillus   
6     f__Lactobacillaceae    g__Ligilactobacillus   
8     f__Lactobacillaceae    g__Ligilactobacillus   
12    f__Lactobacillaceae  g__Limosilactobacillus   
...                   ...                     ...   
2358  f__Lactobacillaceae        g__Lactobacillus   
2359  f__Lactobacillaceae    g__Ligilactobacillus   
2361  f__Lactobacillaceae  g__Limosilactobacillus   
2363  f__Lactobacillaceae    g__Ligilactobacillus   
2369  f__Lactobacillaceae        g__Lactobacillus   

                               Species sample  cluster  
2     s__Limosilactobacillus ingluviei    S10     1039  
3          s__Limosilactobacillus oris    S10     1113  
6          s__Ligilactobacillus agilis    S10     1297  
8      s__Ligilactobacillus salivarius    S10     1752  
12    s__Limosilactobacillus vaginalis    S10      188  
...                                ...    ...      ...  
2358       s__Lactobacillus gallinarum     S9      362  
2359                               s__     S9     3913  
2361  s__Limosilactobacillus reuteri_E     S9      442  
2363     s__Ligilactobacillus aviarius     S9     4469  
2369        s__Lactobacillus johnsonii     S9      569  

[331 rows x 13 columns]

In [575]:
NCBI_Lactobacillus_df.to_csv('data56_related_files/GTDB_NCBI_Lactobacillus.csv', index = False)